From 3cbfc8dcfdb91d49e9987d5ff406c559064c3d4a Mon Sep 17 00:00:00 2001
From: Yanteng Si <siyanteng@loongson.cn>
Date: Mon, 20 May 2024 10:20:21 +0800
Subject: [PATCH 144/314] BACKPORT: DEEPIN: LS7A2000: Add quirk for OHCI device
 rev 0x02

Signed-off-by: Baoqi Zhang <zhangbaoqi@loongson.cn>
Signed-off-by: Yanteng Si <siyanteng@loongson.cn>

Link: https://github.com/deepin-community/kernel/commit/1f39aaf5dfa139d112d5d0c32d347aca8634b6c0
[Mingcong Bai: This patch fixes USB full-speed and low-speed devices on
 ASUS XC-LS3A6M boards, when at least two full-/low-speed devices, and
 one SuperSpeed device are pluged-in.

 Test was conducted on v6.13-rc7 and v6.12.10, with or without AOSC OS
 patchset available at the time. The full-/low-speed devices only work
 with this specific patch applied, or at least one of the two
 full-/low-speed devices would not work.

 With more than one SuperSpeed devices plugged in, the issue reproduces
 more reliably, where at least one of the two full-/low-speed devices
 would not work. With three SuperSpeed devices plugged in, filling all
 five downstream facing ports at the back of the board along with the
 two full-/low-speed devices, the issue reproduces consistently.

 Similar issues have also been reported by mini PC (NUC) users, however
 no device was available for testing at the time of writing.

 The mechanism behind this patch remains unclear as the original patch
 supplied to deepin does not explain. We are working with our colleagues
 at Loongson Technology for a better detailed (and hopefully
 upstream-oriented) patch.]
[Kexy: Resolved minor conflict in drivers/pci/controller/pci-loongson.c]
Signed-off-by: Kexy Biscuit <kexybiscuit@aosc.io>
---
 drivers/pci/controller/pci-loongson.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/drivers/pci/controller/pci-loongson.c b/drivers/pci/controller/pci-loongson.c
index bc630ab8a283..8329a11c2f4e 100644
--- a/drivers/pci/controller/pci-loongson.c
+++ b/drivers/pci/controller/pci-loongson.c
@@ -32,6 +32,7 @@
 #define DEV_LS7A_CONF	0x7a10
 #define DEV_LS7A_GNET	0x7a13
 #define DEV_LS7A_EHCI	0x7a14
+#define DEV_LS7A_OHCI	0x7a24
 #define DEV_LS7A_DC2	0x7a36
 #define DEV_LS7A_HDMI	0x7a37
 
@@ -176,6 +177,13 @@ static void loongson_pci_msi_quirk(struct pci_dev *dev)
 }
 DECLARE_PCI_FIXUP_FINAL(PCI_VENDOR_ID_LOONGSON, DEV_LS7A_PCIE_PORT5, loongson_pci_msi_quirk);
 
+static void loongson_ohci_quirk(struct pci_dev *dev)
+{
+	if (dev->revision == 0x2)
+		dev->resource[0].start += 0x1000;
+}
+DECLARE_PCI_FIXUP_FINAL(PCI_VENDOR_ID_LOONGSON, DEV_LS7A_OHCI, loongson_ohci_quirk);
+
 static struct loongson_pci *pci_bus_to_loongson_pci(struct pci_bus *bus)
 {
 	struct pci_config_window *cfg;
-- 
2.48.1

